### ISSUE TITLE
[REQUEST] Support for PBP file 

### ISSUE THREAD

**Issue by lollo78** [libretro/beetle-psx-libretro#36]
Hi, could you add the support for PBP file (like PCSX_Rearmed)?
This kind of file help a lot to save disk space.
Thx a lot for all your hard work!

**Comment by Zapeth** [comment]
I added experimental support for PBP reading using the implementation in the emulator PCSX-R and the existing code in the core as templates, feel free to test it -> https://github.com/Zapeth/beetle-psx-libretro

However there are some things that should be mentioned:

1) official PBP files probably won't load since their TOC section is most likely encrypted (which btw don't load in PCSX-R and ePSXe either, don't know about PCSX_Rearmed) and decrypting those would require adding quite a bit of code, not sure if that's worth the trouble if most people will be using converted raw images anyway (but if someone is interested how its done -> https://github.com/xdotnano/PSXtract)

2) since it is still unknown how the subchannel data is incorporated in the pbp format you'll need to stick with external .sbi files for games using libcrypt (just use the same filename as the .pbp file)

3) multi-disc games are not supported yet (though this is more an issue of being unsure how to integrate it in the core since it kind of conflicts with the existing disc switch option), some _might_ work being split into single pbp files with the m3u playlist feature but its definitely not guaranteed (nor recommended at this point)

4) last but not least I'd very much appreciate it if someone with the know-how behind the PSX CD loading would look at my additions/changes since this is my first time working on that kind of stuff (specifically the code in the CDAccess_PBP::Read_TOC function)

**Comment by twinaphex** [comment]
That is a very cool addition, and I would definitely like to look at how it's done so we could incorporate it into RetroArch's scanning system too.

Since you mention official PBP files don't load, what should be the procedure or tool I have to use in order to convert CD images to this PBP format so I can expect them to run? I'd like to test this code on some images.

**Comment by lollo78** [comment]
@Zapeth: great news, thanks a lot
@twinaphex: I use PSX2PSP http://appnee.com/psx2psp/

**Comment by Zapeth** [comment]
@twinaphex: I also use PSX2PSP which is moreless a GUI for a tool called popstation (its source is included in the download link that lollo78 provided), though I'm not sure if there is a linux version around.

**Comment by twinaphex** [comment]
I cherry-picked your commits and merged them. The commits looked fine to me.

Regarding getting official PBP images to work - I assume we'd have to drag in libkirk ? I'm fine with that.

**Comment by Zapeth** [comment]
Oh ok, I guess if you're ok with it :P
 
Though I haven't really investigated more in the official files yet so I'm not sure how exactly the decryption process works and if there are any more differences in addition to that.

For example I've heard that the audio tracks from those files are also compressed and/or downsampled in some way as opposed to the 1:1 copies of the audio tracks from the raw images in the converted pbp files, so it could be difficult to figure that out if its true.

**Comment by Awakened0** [comment]
> 3) multi-disc games are not supported yet and it will only load the first disc by default (though this is more an issue of being unsure how to integrate it in the core than not knowing how it works in the pbp format since it kind of conflicts with the existing disc switch option), some might work being split into separate pbp files with the m3u playlist feature but its definitely not guaranteed (nor recommended at this point)

I made a .m3u playlist pointing to disc 1 and 2 of Chrono Cross in single disc .pbp files to see if it'd work, but it just gave me a black screen.

**Comment by Zapeth** [comment]
You mean when you switch discs at the ingame prompt?

Again, the m3u feature is not a guaranteed working workaround for pbp files, thats why I put the emphasis on "might" work.
For the time being I'd recommend sticking with raw images for multi-disc games until the according commit(s) are pushed.

But perhaps you can upload a savefile where the ingame prompt appears, that might at least help with testing since I currently only have one game like that available.

**Comment by Awakened0** [comment]
It's working for me now sometimes. I think I was getting a black screen before because I was holding my fast forward key right after booting, and I have skip BIOS on. I guess it doesn't like that. One time when I loaded it I got a Bad Sector 18 error. One time RetroArch just crashed. One time I got a Visual C++ runtime error. It seems playlists are unstable with .pbps right now.

Of course I'm waiting until you guys get multidisc .pbps to really switch over my collection. I just wanted to see if the .m3u method worked with them.

**Comment by Zapeth** [comment]
Are you talking about disc swapping, general emulation or booting up now?

If its the latter two, it shouldn't actually make a difference whether you load the pbp directly or from a playlist and I've actually never encountered any of the issues that you described.
However I also usually don't use fast forwarding or skip BIOS, not sure if those could be problematic with the way pbp files work but I'll try to look into it somewhere down the line.

**Comment by Awakened0** [comment]
The intermittent crashes or errors only happened on loading a .m3u pointing to .pbp files with skip BIOS on. When it does work, I don't seem to have any problem switching discs with my eject and disc next hotkeys. I don't have any saves at disc change spots to test, but I switched to disc 2, reset from RetroArch's menu, then tried to start a new game and got the "Please Insert Disc 1" prompt, confirming I was on disc 2.

With skip BIOS on I haven't had any crashes or errors when I load the disc 1 pbp directly or if I load a .m3u playlist pointing to uncompressed images. If I turn skip BIOS off, I can't get it to crash anymore when using a playlist to load .pbps, so it's probably related.

**Comment by Zapeth** [comment]
With the recent commits multi-disc pbp files should now be supported (though you should use single files instead of split ones in a playlist).
I haven't really encountered any crashes either (except when implementing stuff ofc) so if you still get any please let me know (and under what conditions like the core settings, the game you used and what steps you did in the game/program before the crash occurred).

**Comment by Awakened0** [comment]
I've only tested Chrono Cross (as a single multi disc pbp file) so far, but I tried booting it multiple times and didn't get any crashes so far with skip BIOS enabled. Opening the tray and swapping discs works just as well as the .m3u method. I'll report back as I test more games if I find any problems.

Thanks for adding pbp support Zapeth and twinaphex! I save about 16 GB with the compression and my PSX folder is much cleaner :D

**Comment by Awakened0** [comment]
I tested 70 games and found three that freeze at the PSX logo screen when skip BIOS is disabled. With it enabled they freeze on a black screen. All of these work fine uncompressed and they are all NTSC-U versions:

Resident Evil - Director's Cut - Dual Shock Edition
Resident Evil 2 - Dual Shock Edition
Tales of Destiny II (Tales of Eternia)

I tried bin/cue dumps I converted to pbp myself with PSX2PSP as well as premade pbps.

Here's a verbose log of when I loaded the ToDII pbp: http://hastebin.com/ukopulubot.tex

**Comment by stevenstrike** [comment]
Getting the same problem with Resident Evil 2 - Dual Shock Edition (tested only with no compression), skip BIOS disabled, no matter if Leon or Claire is set as CD1.

**Comment by Zapeth** [comment]
Yes, skip BIOS option doesn't seem to make a difference here.

I actually also tried the official pbp file for RE2 from the PSN store and it loads there so it could also be a problem with the conversion tool here.
Unfortunately its pretty hard to tell what goes wrong since both images seem to load exactly the same, only that in one case the read calls switch to a different section of the disc at some point (getting it past the BIOS) and in the other one it doesn't.

It would be interesting if someone could test these converted pbp files on a PSP to see if they work there (not in RA but in whatever they use PBP files for there), if they don't then its probably a problem with the conversion tool.

**Comment by stevenstrike** [comment]
Resident Evil 2 - Dual Shock Edition + PSP1000 using 6.61 ME2.3 CFW (this time i set PBP compression to max) : Works fine

**Comment by Zapeth** [comment]
I think I managed to fix it, at least RE and RE2 boot now.

Just need to wait until the commits are pushed upstream (https://github.com/Zapeth/beetle-psx-libretro/commit/8155cfc641f65a421e1680b1fc68695ff7f8d3a0 if you dont want to wait)

**Comment by stevenstrike** [comment]
So i managed to setup a toolchain to compile your fork, so this is what i tested :
Load PBP (disc 1 is leon) -> Now Boots, everything is fine, leon stuff starts
Goto Retroarch menu -> set to disc 2 (claire here) -> restart -> Same, claire stuff starts

TLDR : Re2DS works, changing cd's works. Good job @Zapeth :+1:

**Comment by Awakened0** [comment]
I can confirm both Resident Evil games now working as well as Destiny II / Eternia. Disc switching with them seems all good too.

Awesome, quick work! Thanks again Zapeth!

**Comment by Zapeth** [comment]
Nice, then I guess this issue can be closed until someone stumbles upon a game that shows some problems with the format.

However one last thing that I probably should mention is that in case of games converted with PSX2PSP, I noticed that the offsets/sizes for some of the audio tracks are slightly off (example: The Italian Job (PAL), menu music).
I do think that this is a problem of the conversion tool since there isn't much room for variation with whats written in the file (not to mention that other emulators act the same way with the converted game).

Unfortunately the source of the application is not available (only the source for the popstation module which doesn't include the code for generating the TOC data) so I can't look into it but since the game should still be perfectly playable I guess its just a minor issue.

If anyone knows a better alternative for a conversion tool (preferably open source) feel free to let me know :)

**Comment by stevenstrike** [comment]
> 
If anyone knows a better alternative for a conversion tool (preferably open source) feel free to let me know :)

I just found this little piece of software : [ipops](https://ipops.julianxhokaxhiu.com/), [Github](https://github.com/julianxhokaxhiu/iPoPS), never heard of it (part of why is that i never owned a ppc/intel mac), could be interesting (for science i guess)

Funny thing is that this git contains a 'base.pbp' which is not a legal file to redistribute.

**Comment by Zapeth** [comment]
> (part of why is that i never owned a ppc/intel mac), could be interesting (for science i guess)
Me neither, and I guess the effort for porting that might as well be spent on making a tool from scratch (something that I don't feel like doing).

But thanks anyway for the reference, someone might be able to do something with this.

**Comment by dankcushions** [comment]
I wonder if it's possible to get multidisc PBP support working with a single, combined file? I think programmatically it's possible, but I'm not sure how best to have a disc selector option to the retroarch UI. I think you can build up core options at game load, so maybe possible there?

**Comment by lollo78** [comment]
@twinaphex 
Hi, retroarch cannot scan PBP files.
There is a way to fix this?
Thx

**Comment by Awakened0** [comment]
> I wonder if it's possible to get multidisc PBP support working with a single, combined file? I think programmatically it's possible, but I'm not sure how best to have a disc selector option to the retroarch UI. I think you can build up core options at game load, so maybe possible there?

It already does. It works like the .m3u playlist support. You load a single file multidisc PBP, then use RetroArch's disc eject, disc next and disc previous hotkeys to navigate. There is also a Disc Control entry in the Quick Menu when you load a multidisc game where you can change the disc index and open or shut the tray.

**Comment by tony971** [comment]
> It already does. It works like the .m3u playlist support. You load a single file multidisc PBP, then use RetroArch's disc eject, disc next and disc previous hotkeys to navigate. There is also a Disc Control entry in the Quick Menu when you load a multidisc game where you can change the disc index and open or shut the tray.

This should probably be documented in README.md.

**Comment by rz5** [comment]
@tony971 Write what you think is appropriate and I'll modify it. Or make a pull request. I'll pull the hell out of it.

**Comment by twinaphex** [comment]
I think we can close this now, we have implemented this to the best ability we could.

**Comment by bigboo3000** [comment]
Hi, I think I found a problem with your pbp reading code:
https://github.com/libretro/beetle-psx-libretro/issues/130
